plugins/trace_api_plugin/compressed_file.cpp

```diff
@@ -0,0 +1,311 @@
+}
+
+compressed_file::~compressed_file()
+{}
+
+void compressed_file::seek( long loc ) {
+   impl->seek(loc, *file_ptr);
+
+}
+
+void compressed_file::read( char* d, size_t n ) {
+   impl->read(d, n, *file_ptr);
+}
+
+// these are defaulted now that the opaque impl type is known
+//
+compressed_file::compressed_file( compressed_file&& ) = default;
+compressed_file& compressed_file::operator= ( compressed_file&& ) = default;
+
+
+/**
+ * Create a compressed file that looks like this:
+ *
+ * /====================\ file offset 0
+ * |                    |
+ * |  Compressed Data   |
+ * |  with seek points  |
+ * |                    |
+ * |--------------------|  file offset END - 2 - (16 * seek point count)
+ * |                    |
+ * |  mapping of        |
+ * |    orig offset to  |
+ * |    seek pt offset  |
+ * |                    |
+ * |--------------------|  file offset END - 2
+ * |  seek pt count     |
+ * \====================/  file offset END
+ *
+ * Where a "seek point" is a point in the compressed data stream where
+ * the decompressor can start reading from having not read any of the prior data
+ * seek points should be traversable by a decompressor so that reads which span
+ * seek points do not have to be aware of them
+ *
+ * In zlib this is created by doing a complete flush of the stream
+ *
+ * @param input_path
+ * @param output_path
+ * @param seek_point_count
+ * @return
+ */
+bool compressed_file::process( const fc::path& input_path, const fc::path& output_path, uint16_t seek_point_count ) {
+   if (!fc::exists(input_path)) {
+      throw std::ios_base::failure(std::string("Attempting to create compressed_file from file that does not exist: ") + input_path.generic_string());
+   }
+   std::vector<seek_point_entry> seek_point_map(seek_point_count);
+
+   size_t input_size = fc::file_size(input_path);
+   auto ideal_seek_stride = input_size / seek_point_map.size();
+
+   fc::cfile input_file;
+   input_file.set_file_path(input_path);
+   input_file.open("rb");
+
+   fc::cfile output_file;
+   output_file.set_file_path(output_path);
+   output_file.open("wb");
+
+   z_stream strm;
+   strm.zalloc = Z_NULL;
+   strm.zfree = Z_NULL;
+   strm.opaque = Z_NULL;
+
+   if (deflateInit2(&strm, Z_BEST_COMPRESSION, Z_DEFLATED, -15, 8, Z_DEFAULT_STRATEGY) != Z_OK) {
+      return false;
+   }
+
+   constexpr size_t buffer_size = 64*1024;
+   auto input_buffer = std::vector<uint8_t>(buffer_size);
+   auto output_buffer = std::vector<uint8_t>(buffer_size);
+
+   auto bytes_remaining_before_sync = ideal_seek_stride;
+   int next_sync_point = 0;
+
+   size_t read_offset = 0;
+   while (read_offset < input_size) {
+      auto bytes_remaining = input_size - read_offset;
+      auto read_size = std::min({ buffer_size, bytes_remaining, bytes_remaining_before_sync });
+      input_file.read(reinterpret_cast<char*>(input_buffer.data()), read_size);
+
+      strm.avail_in = read_size;
+      strm.next_in = input_buffer.data();
+
+      do {
```

Comment:
Can we make this a method, or a lambda and pass in Z_NO_FLUSH/Z_FINISH/Z_FULL_FLUSH?  Should method assert str.avail_in == 0 for Z_NO_FLUSH? (or set next_in to input_buffer and set avail_in also)